horovod/common/operations.cc

```diff
@@ -0,0 +1,1577 @@
+              << " tensor shapes: One rank sent a tensor with dimension " << dim
+              << " equal to " << tensor_shape.dim_size(dim)
+              << ", but another rank sent a tensor with dimension " << dim
+              << " equal to " << request_shape.dim_size(dim) << ".";
+          dim_mismatch = true;
+          break;
+        }
+      }
+      if (dim_mismatch) {
+        break;
+      }
+
+      tensor_sizes[requests[i].request_rank()] = request_shape.dim_size(0);
+    }
+  }
+
+  // If we are doing a broadcast, check that all root ranks are identical.
+  if (message_type == MPIRequest::BROADCAST) {
+    int first_root_rank = requests[0].root_rank();
+    for (unsigned int i = 1; i < requests.size(); i++) {
+      if (error) {
+        break;
+      }
+
+      int this_root_rank = requests[i].root_rank();
+      if (first_root_rank != this_root_rank) {
+        error = true;
+        error_message_stream
+            << "Mismatched " << MPIRequest::RequestType_Name(message_type)
+            << " root ranks: One rank specified root rank " << first_root_rank
+            << ", but another rank specified root rank " << this_root_rank
+            << ".";
+        break;
+      }
+    }
+  }
+
+  bool first_device_is_cpu = requests[0].device() == CPU_DEVICE_ID;
+  for (unsigned int i = 1; i < requests.size(); i++) {
+    if (error) {
+      break;
+    }
+
+    bool this_device_is_cpu = requests[i].device() == CPU_DEVICE_ID;
+    if (first_device_is_cpu != this_device_is_cpu) {
+      error = true;
+      error_message_stream
+          << "Mismatched " << MPIRequest::RequestType_Name(message_type)
+          << " CPU/GPU device selection: One rank specified device "
+          << (first_device_is_cpu ? "CPU" : "GPU")
+          << ", but another rank specified device "
+          << (this_device_is_cpu ? "CPU" : "GPU") << ".";
+      break;
+    }
+  }
+  std::vector<int32_t> devices(requests.size());
+  for (auto it = requests.begin(); it != requests.end(); it++) {
+    devices[it->request_rank()] = it->device();
+  }
+
+  MPIResponse response;
+  response.add_tensor_names(name);
+  if (error) {
+    std::string error_message = error_message_stream.str();
+    response.set_response_type(MPIResponse::ERROR);
+    response.set_error_message(error_message);
+  } else if (message_type == MPIRequest::ALLGATHER) {
+    response.set_response_type(MPIResponse::ALLGATHER);
+    for (auto dim : tensor_sizes) {
+      response.add_tensor_sizes(dim);
+    }
+  } else if (message_type == MPIRequest::ALLREDUCE) {
+    response.set_response_type(MPIResponse::ALLREDUCE);
+  } else if (message_type == MPIRequest::BROADCAST) {
+    response.set_response_type(MPIResponse::BROADCAST);
+  }
+  response.set_devices(devices);
+
+  // Clear all queued up requests for this name. They are now taken care of
+  // by the constructed MPI response.
+  message_table->erase(it);
+
+  return response;
```

Comment:
A valid C++ pattern is to have like this:

```
return response;

error_return:
    return Failure;
```

and then instead of all of these early exit `if error` you can just do `goto error_return;`